### ISSUE TITLE
JIT doesn't inline methods contain EH code

### ISSUE THREAD

**Issue by hez2010** [dotnet/runtime#62038]
### Description

JIT doesn't inline methods contain EH code.

Let's say this code:

```csharp
void Test()
{
	Console.WriteLine(Foo(3));
}

int Foo(int x)
{
	try
	{
		return Get(x);
	}
	catch
	{
		return 0;
	}
}

int Get(int x) => x is 1 ? throw new Exception() : x + 1;
```

Expected codegen for `Test`:
```asm
mov	ecx, 4
jmp	System.Console.WriteLine(Int32)
```

Actual codegen for `Test`:
```asm
sub	rsp, 0x28
mov	ecx, 3
call	Program.Foo(Int32)
mov	ecx, eax
add	rsp, 0x28
jmp	System.Console.WriteLine(Int32)
```

Note that even I added `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on top of the method `Foo`, it won't be inlined either.

### Configuration

.NET 6

**Comment by AndyAyersMS** [comment]
This is a long-standing limitation in the jit, and non-trivial to fix.

If you know of examples in existing code bases where you think enabling this would be helpful, please point them out.

**Comment by MichalPetryka** [comment]
> If you know of examples in existing code bases where you think enabling this would be helpful, please point them out.

The most impactful cases for this would probably be `using`s and `foreach`es (the latter would maybe then allow their devirtualization in more cases which would help stuff like LINQ).

**Comment by AndyAyersMS** [comment]
Status update: partially fixed.

Verified on current main with `JitPrintInlinedMethods`:

| variant | inlining result |
| --- | --- |
| `try`/`finally` | `[INLINED: call site: profitable inline]` ✅ |
| `try`/`catch` (the original repro) | `[FAILED: callee: has exception handling]` ❌ |

The general EH-inlining work landed in @jakobbotsch''s #108900 (see PRs #112998, #113023, etc.), which covers `try`/`finally`/`fault`/`filter`. The `using` and `foreach` patterns @MichalPetryka called out are all `try`/`finally`, so the most-requested cases are now inlined.

What''s still blocked is **catch-clause inlining** specifically -- gated in `src/coreclr/jit/fgbasic.cpp:3527-3532`:

```cpp
if (!isFinallyFaultOrFilter)
{
    JITDUMP("Inlinee EH clause %u is a catch; we can''t inline these (yet)\n", XTnum);
    compInlineResult->NoteFatal(InlineObservation::CALLEE_HAS_EH);
    return;
}
```

The comment above that block notes the reason -- catch class is represented by a token and we don''t currently have a way to convey that across inline boundaries -- and a possible path via runtime filters along the lines of `fgCreateFiltersForGenericExceptions`.

Leaving this open as a tracking item for that remaining catch case.

**Comment by jakobbotsch** [comment]
> The general EH-inlining work landed in [@jakobbotsch](https://github.com/jakobbotsch)''s [#108900](https://github.com/dotnet/runtime/issues/108900) (see PRs [#112998](https://github.com/dotnet/runtime/pull/112998), [#113023](https://github.com/dotnet/runtime/pull/113023), etc.), which covers `try`/`finally`/`fault`/`filter`

Couldn't have done it without you @AndyAyersMS 😉

**Comment by jkotas** [comment]
> catch class is represented by a token and we don't currently have a way to convey that across inline boundaries -- and a possible path via runtime filters along the lines of fgCreateFiltersForGenericExceptions

We should not need to go all the way to produce filters. The runtime is able to store type handle instead of token in the catch clause today. We use that for dynamic methods. We can start doing that inlinees too.

**Comment by AndyAyersMS** [comment]
> > catch class is represented by a token and we don't currently have a way to convey that across inline boundaries -- and a possible path via runtime filters along the lines of fgCreateFiltersForGenericExceptions
> 
> We should not need to go all the way to produce filters. The runtime is able to store type handle instead of token in the catch clause today. We use that for dynamic methods. We can start doing that inlinees too.

Yes, I saw we had that for LCG methods, so that's what I was thinking of doing for catches. Once we do that, is there any benefit to keeping the token representation, or should we just switch everything over?

**Comment by jkotas** [comment]
My guess is that the token representation was used as a startup performance optimization. It delays resolving the token until exception is thrown and the catch clause is executed.

Also, R2R uses the token representation to persist the EH clause in the binary. If we switched away from tokens at JIT/EE interface level, we would have to recreate the token (this should not be hard) or modify the format.
